core/src/main/java/org/infinispan/protostream/annotations/impl/AbstractMarshallerCodeGenerator.java

```diff
@@ -288,34 +288,67 @@ protected void generateReadMethodBody(IndentWriter iw, ProtoMessageTypeMetadata
             iw.println(";");
          }
       }
-      iw.println("boolean done = false;");
-      iw.println("while (!done) {");
-      iw.inc();
-      iw.println("final int tag = $in.readTag();");
-      iw.println("switch (tag) {");
-      iw.inc();
-      iw.println("case 0: {");
-      iw.inc();
-      iw.println("done = true;");
-      iw.println("break;");
-      iw.dec();
-      iw.println("}");
-      for (ProtoFieldMetadata fieldMetadata : messageTypeMetadata.getFields().values()) {
-         generateFieldReadMethod(messageTypeMetadata, fieldMetadata, iw, noFactory, trackedFields, getUnknownFieldSetFieldStatement, setUnknownFieldSetFieldStatement);
-      }
-      iw.println("default: {");
-      iw.inc();
-      if (getUnknownFieldSetFieldStatement != null) {
-         iw.printf("%s.UnknownFieldSet u = %s;\n", PROTOSTREAM_PACKAGE, getUnknownFieldSetFieldStatement);
-         iw.printf("if (u == null) u = new %s.impl.UnknownFieldSetImpl();\n", PROTOSTREAM_PACKAGE);
-         iw.println("if (!u.readSingleField(tag, $in)) done = true;");
-         iw.printf("if (!u.isEmpty()) %s;\n", setUnknownFieldSetFieldStatement);
+      if (messageTypeMetadata.isOrderedMarshallable()) {
```

Comment:
I really wish we used some kind of templating engine instead of doing this in the code.